Create rule S6380[terraform]: Authorizing anonymous access to Azure resources is security-sensitive (#574)
SonarSource/rspec: rules/S6380/terraform/rule.adoc

> Disabling authentication can reduce an organization's ability to protect itself against attacks on its critical functions and data, if any.

Reviewer: ```suggestion
Authorizing anonymous access can reduce an organization's ability to protect itself against attacks on its critical functions and data, if any.
```

> 

Reviewer: The first sentence should be followed by at least one illustration of the risk/impact:
- Anonymous traffic generating performances issues/ unwanted costs
- Public access to sensitive data

Depending on the affected Azure resource, multiple choices are possible: Active Directory Authentication, OpenID, or numerous other Oauth 2.0 providers.

Authentication Coupled with fine-grained authorizations helps follow the Defense-In-Depth principle and bring traceability to investigators of security incidents.

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.
* Compliance policies require access to this resource to be authenticated

> 

Reviewer: What about "This resource stores or processes sensitive data."

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

> Enable authentication in this Azure resource.

Reviewer: Another recommendation should be to enforce authentication of all access (deny anonymous access)

  }
}
----

== Compliant Solution

For https://azure.microsoft.com/en-us/services/app-service/[App Service authentication]:

----
resource "azurerm_windows_web_app" "example" {
  name = "example"

  auth_settings {
    enabled = true
  }
}
----

For https://docs.microsoft.com/en-us/azure/api-management/api-management-howto-protect-backend-with-aad[OpenID Authentication]:

----
resource "azurerm_api_management_api" "example" {
  name = "example-api"

  openid_authentication {
    openid_provider_name = azurerm_api_management_openid_connect_provider.example.name
    bearer_token_sending_methods = "authorizationHeader"
  }
}
----

For https://azure.microsoft.com/en-us/services/cache/[Redis Caches]:

----
resource "azurerm_redis_cache" "example" {
  name = "example-cache"
  
  redis_configuration {
    enable_authentication = true
  }
}
----

== See

* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Boken Access Control
* https://cwe.mitre.org/data/definitions/668.html[MITRE, CWE-668] - Exposure of Resource to Wrong Sphere

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> Make sure that disabling authentication is safe here.

Reviewer 2: I'm not sure about this one, since parameters differ a lot from each other. Is it possible ?

Disabling authentication can reduce an organization's ability to protect itself against attacks on its critical functions and data, if any.

> Depending on the Azure resource, multiple choices are possible: Active Directory Authentication, TLS mutual authentication or OpenID.

Reviewer 2: ```suggestion
Depending on the Azure resource, multiple choices are possible: Active Directory Authentication, OpenID or multiple other Oauth 2.0 providers.
```

    enabled = true
  }
}
----

For https://docs.microsoft.com/en-us/azure/api-management/api-management-howto-protect-backend-with-aad[OpenID Authentication]:

----
resource "azurerm_api_management_api" "example" {
  name = "example-api"

  openid_authentication {
    openid_provider_name = azurerm_api_management_openid_connect_provider.example.name
    bearer_token_sending_methods = "authorizationHeader"
  }
}
----

For https://azure.microsoft.com/en-us/services/cache/[Redis Caches]:

----
resource "azurerm_redis_cache" "example" {
  name = "example-cache"
  
  redis_configuration {
    enable_authentication = true
  }
}
----

== See

* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Boken Access Control
* https://cwe.mitre.org/data/definitions/668.html[MITRE, CWE-668] - Exposure of Resource to Wrong Sphere

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

Make sure that disabling authentication is safe here.

=== Highlighting

* For ``auth_settings`` cases: Highlight the resource or ``auth_settings->enabled = false``
* For ``api_management_api``: Highlight the resource

> * For ``storage_account``: Highlight the resource

Reviewer 2: ```suggestion
```

Disabling authentication can reduce an organization's ability to protect itself against attacks on its critical functions and data, if any.

Depending on the Azure resource, multiple choices are possible: Active Directory Authentication, TLS mutual authentication or OpenID.

Authentication Coupled with fine-grained authorizations helps follow the Defense-In-Depth principle and bring traceability to investigators of security incidents.

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.
* Compliance policies require access to this resource to be authenticated

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

Enable authentication in this Azure resource.

== Sensitive Code Example

For https://azure.microsoft.com/en-us/services/active-directory[Active Directory Authentication]:

----
resource "azurerm_windows_web_app" "example" {

>   name                = "example"

Reviewer 2: ```suggestion
  name = "example"
```

  auth_settings {
    enabled = false # Sensitive
  }
}
----

For https://docs.microsoft.com/en-us/azure/api-management/api-management-howto-protect-backend-with-aad[OpenID Authentication]:

----
resource "azurerm_api_management_api" "example" { # Sensitive, the openid_authentication block is missing
  name = "example-api"
}
----

> 
> For https://docs.microsoft.com/en-us/azure/app-service/app-service-web-configure-tls-mutual-auth[TLS Mutual Authentication]:
> 
> ----
> resource "azurerm_linux_web_app" "example" { # Sensitive, TLS mutual authentication disabled and set to "Optional"
>   client_cert_enabled = false
>   client_cert_mode = "Optional"
> }
> ----

Reviewer 2: ```suggestion
```

For https://docs.microsoft.com/en-us/azure/storage/files/storage-files-active-directory-overview[Azure Files Authentication]:

----
resource "azurerm_storage_account" "example" { # Sensitive, the azure_files_authentication block is missing
  name = "storageaccountname"
}
----

== Compliant Solution

For https://azure.microsoft.com/en-us/services/active-directory[Active Directory Authentication]:

----
resource "azurerm_windows_web_app" "example" {
  name                = "example"

  auth_settings {
    enabled = true
  }
}
----

For https://docs.microsoft.com/en-us/azure/api-management/api-management-howto-protect-backend-with-aad[OpenID Authentication]:

----
resource "azurerm_api_management_api" "example" {
  name = "example-api"
  openid_authentication {
    openid_provider_name = "google"
    bearer_token_sending_methods = "authorizationHeader"
  }
}
----

> 
> For https://docs.microsoft.com/en-us/azure/app-service/app-service-web-configure-tls-mutual-auth[TLS Mutual Authentication]:
> 
> ----
> resource "azurerm_linux_web_app" "example" {
>   client_cert_enabled = true
>   client_cert_mode = "Required"
> }
> ----

Reviewer 2: ```suggestion
```

Using authentication coupled with fine-grained authorizations helps bring Defense-In-Depth and bring traceability to investigators of security incidents.

Depending on the affected Azure resource, multiple authentication choices are possible: Active Directory Authentication, OpenID implementations (Google, Microsoft, etc.) or native Azure mechanisms.
Unfortunately, specific Azure resources only expose basic authentication, but it should also be an acceptable first layer of defense.

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.
* This Azure resource stores or processes sensitive data.
* Compliance policies require access to this resource to be authenticated.

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

Enable authentication in this Azure resource, and disable anonymous access.

== Sensitive Code Example

For https://azure.microsoft.com/en-us/services/app-service/[App Services and equivalent]:

----
resource "azurerm_function_app" "example" { # Sensitive
  name = "example"

>   auth_settings {
>     enabled = false
>     unauthenticated_client_action = "AllowAnonymous"

Reviewer: ```suggestion
  auth_settings {
    enabled = false # Sensitive
  }
  
  auth_settings {
    enabled = true
    unauthenticated_client_action = "AllowAnonymous" # Sensitive
```

Do the same in "Compliant"

Reviewer 2: done in a commit

----
resource "azurerm_data_factory_linked_service_sftp" "example" {
  authentication_type = "Basic"
  username            = local.creds.username
  password            = local.creds.password
}
----

For https://azure.microsoft.com/en-us/product-categories/storage/[Storage Accounts]:

----
resource "azurerm_storage_account" "example" {
  allow_blob_public_access = true
}

resource "azurerm_storage_container" "example" {
  container_access_type = "private"
}
----

For https://azure.microsoft.com/en-us/services/cache/[Redis Caches]:

----
resource "azurerm_redis_cache" "example" {
  name = "example-cache"
  
  redis_configuration {
    enable_authentication = true
  }
}
----

== See

* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Boken Access Control
* https://owasp.org/www-project-top-ten/2017/A5_2017-Broken_Access_Control[OWASP Top 10 2017 Category A5] - Broken Access Control
* https://cwe.mitre.org/data/definitions/668.html[MITRE, CWE-668] - Exposure of Resource to Wrong Sphere

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> Make sure that fully or partially disabling authentication is safe here.

Reviewer: I am not sure what "partially" disabled mean.

> 

Reviewer: ```suggestion
Omitting {property name} disables authentication. Make sure it is safe here.
```

Reviewer 2: I didn't know that this was possible, I add it

Reviewer 2: As said in a previous comment of mine, this sentence is not suited for every case. I'm not sure why we should add it here. In some cases, authentication is explicitly disabled

Reviewer: I think we should specify the message case by case. It could be either:
* Make sure that disabling authentication is safe here.
* Omitting {property name} disables authentication. Make sure it is safe here.
* Make sure that authorizing anonymous access is safe here.
* Omitting {property name} authorize anonymous access. Make sure it is safe here.

    password = local.creds.password
  }
}
----

For https://azure.microsoft.com/en-us/product-categories/storage/[Storage Accounts]:

----
resource "azurerm_storage_account" "example" {
  allow_blob_public_access = true
}

resource "azurerm_storage_container" "example" {
  container_access_type = "private"
}
----

For https://azure.microsoft.com/en-us/services/cache/[Redis Caches]:

----
resource "azurerm_redis_cache" "example" {
  name = "example-cache"
 
  redis_configuration {
    enable_authentication = true
  }
}
----

== See

* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Boken Access Control
* https://owasp.org/www-project-top-ten/2017/A5_2017-Broken_Access_Control[OWASP Top 10 2017 Category A5] - Broken Access Control
* https://cwe.mitre.org/data/definitions/668.html[MITRE, CWE-668] - Exposure of Resource to Wrong Sphere

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

Preliminary Note: App Service and equivalents resources:

* ``app_service``
* ``app_service_slot``
* ``function_app``
* ``function_app_slot``
* ``linux_web_app``
* ``windows_web_app``

> 

Reviewer 2: ```suggestion

'''

```

* For ``redis_cache``: Make sure that disabling authentication is safe here.
* For ``storage_account``: Make sure that authorizing potential anonymous access is safe here.
* For ``storage_container``: Make sure that authorizing potential anonymous access is safe here.

=== Highlighting

* For App Service and equivalents:
** Highlight the resource if the ``auth_settings`` block is missing
** Highlight ``auth_settings->enabled = false`` regardless of ``auth_settings->unauthenticated_client_action``
** Highlight ``auth_settings->unauthenticated_client_action = "AllowAnonymous"``
* For ``api_management_api``: Highlight the resource if the ``openid_authentication`` block is missing
* For ``api_management``:
** Highlight the resource if the ``sign_in`` block is missing
** Highlight ``sign_in->enabled = false``
* For ``data_factory_linked_service_odata``: Highlight the resource if the ``basic_authentication`` block is missing
* For ``data_factory_linked_service_sftp`` and ``data_factory_linked_service_web``: Highlight ``authentication_type = "Anonymous"``

> * For ``redis_cache``: Highlight ``enable_authentication = false``

Reviewer 2: ```suggestion
* For ``redis_cache``: Highlight ``redis_configuration->enable_authentication = false``
```

Authorizing anonymous access can reduce an organization's ability to protect itself against attacks on its Azure resources.

Security incidents may include disrupting critical functions, data theft, and additional Azure subscription costs due to resource overload.

Using authentication coupled with fine-grained authorizations helps bring Defense-In-Depth and bring traceability to investigators of security incidents.

Depending on the affected Azure resource, multiple authentication choices are possible: Active Directory Authentication, OpenID implementations (Google, Microsoft, etc.) or native Azure mechanisms.

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.
* This Azure resource stores or processes sensitive data.
* Compliance policies require access to this resource to be authenticated.

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

Enable authentication in this Azure resource, and disable anonymous access.

> If only Basic Authentication is available, enable it too.

Reviewer 2: ```suggestion
If only Basic Authentication is available, enable it.
```

  username            = local.creds.username
  password            = local.creds.password
}

resource "data_factory_linked_service_odata" "example" {
  basic_auth {
    username = local.creds.username
    password = local.creds.password
  }
}
----

For https://azure.microsoft.com/en-us/product-categories/storage/[Storage Accounts]:

----
resource "azurerm_storage_account" "example" {
  allow_blob_public_access = true
}

resource "azurerm_storage_container" "example" {
  container_access_type = "private"
}
----

For https://azure.microsoft.com/en-us/services/cache/[Redis Caches]:

----
resource "azurerm_redis_cache" "example" {
  name = "example-cache"
 
  redis_configuration {
    enable_authentication = true
  }
}
----

== See

* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Boken Access Control
* https://owasp.org/www-project-top-ten/2017/A5_2017-Broken_Access_Control[OWASP Top 10 2017 Category A5] - Broken Access Control
* https://cwe.mitre.org/data/definitions/668.html[MITRE, CWE-668] - Exposure of Resource to Wrong Sphere

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

> 
> Preliminary Note: App Service and equivalents resources:
> 
> * ``app_service``
> * ``app_service_slot``
> * ``function_app``
> * ``function_app_slot``
> * ``linux_web_app``
> * ``windows_web_app``
> 
> '''
> 
> * For App Service and equivalent resources:
> ** If ``auth_settings`` block is missing: Omitting ``auth_settings`` disables authentication. Make sure it is safe here.
> ** if ``auth_settings->enabled = false``: Make sure that disabling authentication is safe here.
> ** if ``auth_settings->unauthenticated_client_action = "AllowAnonymous"``: Make sure that authorizing anonymous access is safe here.
> * For ``api_management_api``: Omitting ``openid_authentication`` disables authentication. Make sure it is safe here.
> * For ``api_management`` resources:
> ** If ``sign_in`` block is missing: Omitting ``sign_in`` authorizes anonymous access. Make sure it is safe here.
> ** If ``sign_in->enabled = false``: Make sure that giving anonymous access without enforcing sign in is safe here.
> * For ``data_factory_linked_service_odata``: Omitting the ``basic_authentication`` block disables authentication. Make sure it is safe here.
> * For ``data_factory_linked_service_sftp`` and ``data_factory_linked_service_web``: Make sure that authorizing anonymous access is safe here.
> * For ``redis_cache``: Make sure that disabling authentication is safe here.
> * For ``storage_account``: Make sure that authorizing potential anonymous access is safe here.
> * For ``storage_container``: Make sure that authorizing potential anonymous access is safe here.
> 

Reviewer 2: ```suggestion

* For App Service and equivalent resources:
** If ``auth_settings`` block is missing: Omitting ``auth_settings`` disables authentication. Make sure it is safe here.
** if ``auth_settings->enabled = false``: Make sure that disabling authentication is safe here.
** if ``auth_settings->unauthenticated_client_action = "AllowAnonymous"``: Make sure that authorizing anonymous access is safe here.
* For ``api_management_api``: Omitting ``openid_authentication`` disables authentication. Make sure it is safe here.
* For ``api_management`` resources:
** If ``sign_in`` block is missing: Omitting ``sign_in`` authorizes anonymous access. Make sure it is safe here.
** If ``sign_in->enabled = false``: Make sure that giving anonymous access without enforcing sign-in is safe here.
* For ``data_factory_linked_service_odata``: Omitting the ``basic_authentication`` block disables authentication. Make sure it is safe here.
* For ``data_factory_linked_service_sftp`` and ``data_factory_linked_service_web``: Make sure that authorizing anonymous access is safe here.
* For ``redis_cache``: Make sure that disabling authentication is safe here.
* For ``storage_account``: Make sure that authorizing potential anonymous access is safe here.
* For ``storage_container``: Make sure that authorizing potential anonymous access is safe here.

Note: App Service and equivalents resources:

* ``app_service``
* ``app_service_slot``
* ``function_app``
* ``function_app_slot``
* ``linux_web_app``
* ``windows_web_app``
```

Authorizing anonymous access can reduce an organization's ability to protect itself against attacks on its Azure resources.

Security incidents may include disrupting critical functions, data theft, and additional Azure subscription costs due to resource overload.

Using authentication coupled with fine-grained authorizations helps bring Defense-In-Depth and bring traceability to investigators of security incidents.

Depending on the affected Azure resource, multiple authentication choices are possible: Active Directory Authentication, OpenID implementations (Google, Microsoft, etc.) or native Azure mechanisms.

> Unfortunately, specific Azure resources only expose basic authentication, but it should also be an acceptable first layer of defense.

Reviewer: This sentence is confusing as we don't mentions the resources and we don't explain why basic authentication is not the best choice.
I think it bring confusion and I would rather remove it.

Reviewer 2: This sentence can be confusing for specific developer profiles, yes.

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.
* This Azure resource stores or processes sensitive data.
* Compliance policies require access to this resource to be authenticated.

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

Enable authentication in this Azure resource, and disable anonymous access.

== Sensitive Code Example

For https://azure.microsoft.com/en-us/services/app-service/[App Services and equivalent]:

----
resource "azurerm_function_app" "example" { # Sensitive
  name = "example"

  auth_settings {
    enabled = false

>     unauthenticated_client_action = "AllowAnonymous"

Reviewer 2: Thanks for the feedback. However, I won't agree with your second point

>     unauthenticated_client_action = "AllowAnonymous"

Reviewer: My point here was: if `enabled = false` then `unauthenticated_client_action = "AllowAnonymous"` probably have no impact. Therfore, here I would only highlight `enabled = false`

>     enabled = false
>     unauthenticated_client_action = "AllowAnonymous"

Reviewer 2: Don't you think that we should keep them both at their non-compliant value to avoid confusion?

By adding both of them, my point was to add both non-compliant values instead of 
```
    enabled = true
    unauthenticated_client_action = "AllowAnonymous" # Sensitive
```
or 
```
    enabled = false # Sensitive
    unauthenticated_client_action = "RedirectToLoginPage"
```

In my opinion, it would help the developer understand that both of these parameters are security hotspots and not only one of them. Or we could add ``# Sensitive`` in front of both the parameters to make it even clearer:

```suggestion
    enabled = false # Sensitive
    unauthenticated_client_action = "AllowAnonymous" # Sensitive
```

== Compliant Solution

For https://azure.microsoft.com/en-us/services/app-service/[App Services and equivalent]:

----
resource "azurerm_function_app" "example" {
  name = "example"

  auth_settings {
    enabled = true
    unauthenticated_client_action = "RedirectToLoginPage"
  }
}
----

For https://azure.microsoft.com/en-us/services/api-management/[API Management]:

----

> resource "azurerm_api_management_api" "example" { # Sensitive, the openid_authentication block is missing

Reviewer: ```suggestion
resource "azurerm_api_management_api" "example" {
```
